api/hypershift/v1beta1/hostedcluster_types.go

```diff
@@ -703,76 +743,124 @@ type NodePortPublishingStrategy struct {
+	// This field is immutable.
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf", message="baseDomainPrefix is immutable"
+	// +kubebuilder:validation:XValidation:rule=`self.matches('^(?:[a-zA-Z0-9-]+\\.)+[a-zA-Z]{2,}$')`,message="baseDomainPrefix must be a valid base domain (e.g., example.com)"
+	// +kubebuilder:validation:MaxLength=253
+	// +kubebuilder:validation:MinLength=1
 	// +optional
-	// +immutable
 	BaseDomainPrefix *string `json:"baseDomainPrefix,omitempty"`
 
-	// PublicZoneID is the Hosted Zone ID where all the DNS records that are
-	// publicly accessible to the internet exist.
-	//
+	// publicZoneID is the Hosted Zone ID where all the DNS records that are publicly accessible to the internet exist.
+	// This field is optional and mainly leveraged in cloud environments where the DNS records for the .baseDomain are created by controllers in this zone.
+	// Once set, this value is immutable.
 	// +optional
+	// +kubebuilder:validation:XValidation:rule=`oldSelf == "" || self == oldSelf`, message="clusterID is immutable"
+	// +kubebuilder:validation:MaxLength=253
+	// +kubebuilder:validation:MinLength=1
 	// +immutable
 	PublicZoneID string `json:"publicZoneID,omitempty"`
 
-	// PrivateZoneID is the Hosted Zone ID where all the DNS records that are only
-	// available internally to the cluster exist.
-	//
+	// privateZoneID is the Hosted Zone ID where all the DNS records that are only available internally to the cluster exist.
+	// This field is optional and mainly leveraged in cloud environments where the DNS records for the .baseDomain are created by controllers in this zone.
+	// Once set, this value is immutable.
 	// +optional
+	// +kubebuilder:validation:XValidation:rule=`oldSelf == "" || self == oldSelf`, message="clusterID is immutable"
+	// +kubebuilder:validation:MaxLength=253
+	// +kubebuilder:validation:MinLength=1
 	// +immutable
 	PrivateZoneID string `json:"privateZoneID,omitempty"`
 }
 
-// ClusterNetworking specifies network configuration for a cluster.
+// clusterNetworking specifies network configuration for a cluster.
+// All CIDRs must be unique. Additional validation to check for CIDRs overlap and consistent network stack is perfomed by the controllers.
+// Failing that validation will result in the HostedCluster being degraded and the validConfiguration condition being false.
+// +kubebuilder:validation:XValidation:rule="(!has(self.machineNetwork) && self.clusterNetwork.all(c, self.serviceNetwork.all(s, c.cidr != s.cidr)) || (has(self.machineNetwork) && (self.machineNetwork.all(m, self.clusterNetwork.all(c, m.cidr != c.cidr)) && self.machineNetwork.all(m, self.serviceNetwork.all(s, m.cidr != s.cidr)) && self.clusterNetwork.all(c, self.serviceNetwork.all(s, c.cidr != s.cidr)))))",message="CIDR ranges in machineNetwork, clusterNetwork, and serviceNetwork must be unique and non-overlapping"
```

Comment:
How can you tell if they are unique by comparing them as strings? CIDR ranges can actually be non-unique because of the bit length, `192.168.0.1/24` is not a correct manifestation of a CIDR network, since it has non-zero bits after the bit mask, however, it would equate to `192.168.0.0/24` if you parsed it as a CIDR

I also don't think you can tell if they are overlapping in this way.

Since OCP 4.16 we've had a CIDR library in the API server

You probably want something like:
```
isCIDR(self) && !cidr(self).containsCIDR(other)
```
I think you probably also want to try concatenating the lists to make this easier to deal with
```
(self.?machineNetwork.orValue([]) + self.clusterNetwork + self.serviceNetwork).all(x, 
    (self.?machineNetwork.orValue([]) + self.clusterNetwork + self.serviceNetwork).exists_one(y, cidr(x).containsCIDR(y))
)
```
Validating that they are in fact valid CIDRs in a separate rule is also probably preferable, so that one can fail first, and make this one easier to read, you can do that on the struct field itself though.

Al of the above works from 4.16 onwards

Reply:
nice, does isCIDR support ipv6 format?

Reply:
Seems it does based on https://github.com/openshift/api/blob/a2817b89f7e0989016967055a142d5e88bed18de/operator/v1/types_network.go#L487

Reply:
See https://github.com/openshift/hypershift/pull/5090#discussion_r1836890598